include/deal.II/numerics/cell_data_transfer.templates.h

```diff
@@ -0,0 +1,196 @@
+
+#ifndef dealii_cell_data_transfer_templates_h
+#define dealii_cell_data_transfer_templates_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/numerics/cell_data_transfer.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace
+{
+  template <typename VectorType>
+  void
+  post_unpack_action(VectorType &out)
+  {
+    out.compress(::dealii::VectorOperation::insert);
+  }
+
+  template <typename value_type>
+  void
+  post_unpack_action(std::vector<value_type> &)
+  {
+    // Do nothing for std::vector as VectorType.
+  }
+} // namespace
+
+
+template <int dim, int spacedim, typename VectorType>
+CellDataTransfer<dim, spacedim, VectorType>::CellDataTransfer(
+  const Triangulation<dim, spacedim> &                triangulation,
+  const std::function<value_type(
+    const std::vector<value_type> &children_indices)> coarsening_strategy)
+  : triangulation(&triangulation, typeid(*this).name())
+  , coarsening_strategy(coarsening_strategy)
+{
+  Assert(
+    (dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim> *>(
+       &triangulation) == nullptr),
+    ExcMessage("You are calling the CellDataTransfer class "
+               "with a parallel::distributed::Triangulation. "
+               "You probably want to use the "
+               "parallel::distributed::CellDataTransfer class."));
+
+#ifdef DEBUG
+  n_active_cells_pre = numbers::invalid_unsigned_int;
+#endif
+}
+
+
+
+template <int dim, int spacedim, typename VectorType>
+void
+CellDataTransfer<dim, spacedim, VectorType>::
+  prepare_for_coarsening_and_refinement()
+{
+  // Cleanup previous indices.
+  refined_cells_active_index.clear();
+  coarsened_cells_active_index.clear();
+  persisting_cells_active_index.clear();
+
+  for (const auto &cell : triangulation->active_cell_iterators())
+    {
+      // Store iterator and active cell index of cells that will be refined.
+      if (cell->refine_flag_set())
+        {
+          refined_cells_active_index.insert({cell, cell->active_cell_index()});
+        }
+      else if (cell->coarsen_flag_set())
+        {
+          // Gather the iterator to the parent cell of cells that will be
+          // coarsened. Store it together with the active cell indices of all
+          // its children.
+          Assert(cell->level() > 0, ExcInternalError());
+          const auto &parent = cell->parent();
+
+          // Check if the active_cell_indices for the current cell have
+          // been determined already.
+          if (coarsened_cells_active_index.find(parent) ==
+              coarsened_cells_active_index.end())
+            {
+              std::set<unsigned int> indices_children;
+              for (unsigned int child_index = 0;
+                   child_index < parent->n_children();
+                   ++child_index)
+                {
+                  const auto sibling = parent->child(child_index);
+                  Assert(sibling->active(), ExcInternalError());
```

Comment:
Like my refinement example: could we coarsen more than one level at once? I think we can but perhaps there is some other precondition enforced that I don't know about.